lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -2197,6 +2197,21 @@ bool TypeChecker::typeCheckCondition(Expr *&expr, DeclContext *dc) {
     virtual Expr *appliedSolution(constraints::Solution &solution,
                                   Expr *expr) {
       auto &cs = solution.getConstraintSystem();
+
+      // Since conversion constraint is going to enforce actual coercion
+      // only in some cases e.g. function application, we need to explicitly
+      // coerce expression to boolean type here before proceeding.
+      if (!expr->getType()->isBool()) {
```

Comment:
Ah, sorry, with this comment I meant to say that although conversion constraint is going to affect solution it's not actually going to enforce AST transformation, it does so only in certain cases e.g. function application, closure, assignment or if there is contextual type. But if's tuple it's going to be untouched because conversion constraint is just going to be `tuple-to-scalar` and that's it, so we need to explicitly coerce expression to bool here.